addons/point_of_sale/controllers/main.py

```diff
@@ -84,3 +85,141 @@ def print_sale_details(self, date_start=False, date_stop=False, **kw):
+                    with_context.update({f'default_{field.name}': invoice_values.get(field.name)})
+                # Allowing default values for moves is important for some localizations that would need specific fields to be set on the invoice, such as Mexico.
+                pos_order.with_context(with_context).action_pos_order_invoice()
+                return request.redirect('/my/invoices/%s?access_token=%s' % (pos_order.account_move.id, pos_order.account_move._portal_ensure_token()))
+
+        # Most of the time, the country of the customer will be the same as the order. We can prefill it by default with the country of the company.
+        if 'country_id' not in form_values:
+            form_values['country_id'] = pos_order.company_id.account_fiscal_country_id.id
+
+        return request.render("point_of_sale.ticket_validation_screen", {
+            'partner': request.env['res.partner'],
+            'format_amount': format_amount,
+            'env': request.env,
+            'countries': request.env['res.country'].sudo().search([]),
+            'states': request.env['res.country.state'].sudo().search([]),
+            'partner_can_edit_vat': True,
+            'pos_order': pos_order,
+            'invoice_required_fields': additional_invoice_fields,
+            'partner_required_fields': additional_partner_fields,
+            'order_reference': order_reference,
+            'access_token': access_token,
+            **form_values,
+        })
+
+    def _get_matching_partner(self, partner_vals, company_id):
+        """ Try to find a partner based on a multiple parameters.
+        - If the vat is not provided, try to match on the name and email.
+            - If a record match, use it, else create a new one.
+        - If it is provided, first match on it.
+            - If only one record match, use it.
+            - If no record match, create a new one.
+            - If there is multiple matches, try to match on TVA + name.
+                - If there is a match, reuse the partner and if not, create a new record.
+        """
+        ResPartner = request.env['res.partner']
+        if not partner_vals:
+            return ResPartner
+
+        if not partner_vals.get('vat'):
+            ResPartner.flush_model(['company_id', 'name', 'email'])
+            request._cr.execute(
+                '''
+                    SELECT id
+                    FROM res_partner
+                    WHERE (company_id IS NULL OR company_id = %s)
+                        AND unaccent(name) ILIKE unaccent(%s)
+                        AND unaccent(email) ILIKE unaccent(%s)
+                ''',
+                [company_id.id, partner_vals.get('name'), partner_vals.get('email')],
+            )
+            rows = request._cr.fetchall()
+            if len(rows) == 1:
+                return ResPartner.browse(rows[0][0])
+            else:
+                return ResPartner.sudo().create(partner_vals)
```

Comment:
So if you already have 2 partners with the same email, you create a third one? Not sure that's a good idea for the long term :)

Reply:
These heuristics for matching an existing partner seem too risky. Anyone can obtain a valid QR code and then provide arbitrary info when asking for an invoice. VAT numbers are not secrets, and neither are names (they're not even unique).

We can't update partners in the database just because they *look similar* to what the user is providing.

Short of having a clear trusted source of information, or doing identity verification first (e.g. email verification) - we can't do that, we have to create a new partner.

It's exactly the same thing as doing guest purchases on the eCommerce - we can never automatically merge partners unless we have a proof of their identity.

Reply:
BTW, I could not find the code that would update the missing values on the existing partner, when matching one? I suppose that was the planned behavior, otherwise the *partner_localisation_fields_required_to_invoice* may still be missing, no?

Anyway that leads the next can of worms: what values do we allow updating an the existing partner, since we can't verify the identity of the person providing those values?
So really, the only safe course of action is to always create new partners in this context, so we eliminate this problem.

If we want to allow an existing partner to be reused, like on the eCommerce, we can offer customers the option to login in order to skip the form.

Reply:
We are just as concerned as you are about this flow. Initially, I wanted to require the user to be connected so that he can edit his data easily on the portal form but it was decided to not do that.

I talked about it with @chklop and decided that we should never update the user in this flow, as it's not safe. I think the conclusion was that worst case scenario if the data is wrong the user can contact the company to correct them and redo the invoice.

Creating a new partner each time would indeed be the safest but then there is the risk of duplicating a partner multiple times on the db. (Although, if that customer needs his invoice regularly he should ask them at the time of checkout and not always later from this route)

Reply:
alright, but keep in mind that we're still allowing unverified modifications of the partner to some extent, because creating an invoice is a significant operation on that partner.

Moreover, the behavior seems undefined when a matching partner exists but has required info missing? Even if the user provides it in the form (as they will be asked to), the data will be ignored and the creation of the invoice could fail.

I would then recommend:
- add very clear comments in the code to indicate that we *purposely* never update partners when not creating them, and why (it currently looks like an omission)
- catch all possible errors during the creation of the invoice, log them for troubleshooting, and fail gracefully by presenting a clear message to the user, e.g. "Your invoice can't be created automatically at this time, please contact the store to request it"
- cover this scenario in the tests, to make sure this is clearly established as the expected behavior. This way there won't be a bugfix in 2 months that will try to "fix" the problematic cases